lib/pbench/agent/tool_meister.py

```diff
@@ -273,63 +286,69 @@ def install(self):
+        assert self.tool_dir is None, f"tool directory already defined: {self.tool_dir}"
+        assert (
+            self.start_process is None
+        ), f"Tool({self.name}) has an unexpected start process running"
+        assert (
+            self.stop_process is None
+        ), f"Tool({self.name}) has an unexpected stop process running"
 
         args = [
-            f"{self.pbench_install_dir}/tool-scripts/{self.name}",
+            self.tool_script,
             "--start",
-            f"--dir={self.tool_dir}",
+            f"--dir={tool_dir}",
             self.tool_opts,
         ]
         self.logger.info("%s: start_tool -- %s", self.name, " ".join(args))
-        self.start_process = self._create_process_with_logger(
-            args, self.tool_dir, "start"
-        )
+        self.start_process = self._create_process_with_logger(args, tool_dir, "start")
+        self.tool_dir = tool_dir
 
     def stop(self):
         """Stops the background process by running the tool's "stop" operation."""
-        assert self.tool_dir is not None, "Logic bomb!  no tool directory provided!"
-        if not self.tool_dir.is_dir():
-            raise RuntimeError(f"tool directory does not exist: {self.tool_dir}")
-        if self.start_process is None:
-            raise ToolException(f"Tool({self.name})'s start process not running")
-        if self.stop_process is not None:
-            raise ToolException(
-                f"Tool({self.name}) has an unexpected stop process running"
-            )
+        assert self.tool_dir is not None, f"tool directory not defined: {self.tool_dir}"
+        assert (
+            self.start_process is not None
+        ), f"Tool({self.name})'s start process not running"
+        assert (
+            self.stop_process is None
+        ), f"Tool({self.name}) has an unexpected stop process running"
 
         # Before we "stop" a tool, check to see if a "{tool}/{tool}.pid" file
         # exists.  If it doesn't, wait for a second for it to show up.  If
         # after a second it does not show up, then give up waiting and just call
         # the stop method.
         tool_pid_file = self.tool_dir / self.name / f"{self.name}.pid"
-        cnt = 0
-        while not tool_pid_file.exists() and cnt < 100:
+        cnt = 100
+        found = False
+        while cnt > 0 and not found:
+            cnt -= 1
+            if tool_pid_file.exists():
+                found = True
             time.sleep(0.1)
-            cnt += 1
-        if not tool_pid_file.exists():
+        if not found:
             self.logger.warning(
                 "Tool(%s) pid file, %s, does not exist after waiting 10 seconds",
                 self.name,
                 tool_pid_file,
             )
 
         args = [
-            f"{self.pbench_install_dir}/tool-scripts/{self.name}",
+            self.tool_script,
             "--stop",
             f"--dir={self.tool_dir}",
             self.tool_opts,
```

Comment:
This is a some-what hacky way to deal with the poor interfaces to the legacy tool scripts.  The process that actually creates the `pid` file is potentially a sub-process of a sub-process, and various scheduling delays can happen with short start/stop periods.

Since the `TransientTool` class does not yet "own" the actual tool processes, we just move on and allow the actual `stop` script to do what it thinks is best.

Reply:
So, then, waiting for the PID file is an optimization -- that is, we wait for a fixed period of time and then proceed, but if the PID file shows up, then we know that we don't have to wait any more?

Reply:
It is not an optimization.  If a tool's "stop" script can't handle being called without a pid file, we orphan tools tool easily.  This is a fragility that needs to be addressed by remove the legacy tool scripts.  It really can't be part of this PR.

Reply:
I grok that there is a lot of history here, and that there is significant fragility here, and that addressing that is probably out of scope for this PR.  However, I'm still trying to understand the implications of the change to the code here.

This code is waiting for the PID file to appear, and then it's proceeding to invoke the "stop" script, regardless of whether the PID file shows up or not.  And, moreover, once the PID file appears, this code doesn't actually do anything with it!  So, at most, from the perspective of this code, the PID file is just a synchronization mechanism.

Is there some part of the interface contract with the stop script that the PID file has to exist before the script is invoked?  If so, then I can see that this code would have to wait for it...but, if it never appears, then calling the stop script anyway is a violation of the contract.  On the other hand, if that's not part of the contract, then this code shouldn't be concerning itself with the PID file -- that should be left to the stop script (if it cares).

Either way, this code looks potentially broken to me.